nav2-usability/scenario-code/nested-routing/router.dart

```diff
@@ -0,0 +1,309 @@
+  @override
+  Future<AppRoutePath> parseRouteInformation(
+      RouteInformation routeInformation) async {
+    var uri = Uri.parse(routeInformation.location!);
+    if (uri.pathSegments.isEmpty) {
+      return NewBooksRoutePath();
+    }
+    if (uri.pathSegments.length == 1) {
+      if (uri.pathSegments[0] == 'books') return NewBooksRoutePath();
+      if (uri.pathSegments[0] == 'settings') return SettingsRoutePath();
+    }
+
+    if (uri.pathSegments.length == 2) {
+      if (uri.pathSegments[1] == 'new') return NewBooksRoutePath();
+      if (uri.pathSegments[1] == 'all') return AllBooksRoutePath();
+    }
+
+    return NewBooksRoutePath();
+  }
+
+  @override
+  RouteInformation restoreRouteInformation(AppRoutePath path) {
+    late final location;
+
+    if (path is NewBooksRoutePath) {
+      location = '/books/new';
+    } else if (path is AllBooksRoutePath) {
+      location = '/books/all';
+    } else if (path is SettingsRoutePath) {
+      location = '/settings';
+    }
+
+    return RouteInformation(location: location);
+  }
+}
+
+class BookRouterDelegate extends RouterDelegate<AppRoutePath>
+    with ChangeNotifier, PopNavigatorRouterDelegateMixin<AppRoutePath> {
+  @override
+  final GlobalKey<NavigatorState> navigatorKey;
+  final AppState _appState = AppState();
+
+  BookRouterDelegate() : navigatorKey = GlobalKey<NavigatorState>() {
+    _appState.addListener(() => notifyListeners());
+  }
+
+  @override
+  Future<void> setNewRoutePath(AppRoutePath path) async {
+    if (path is NewBooksRoutePath) {
+      _appState.page = AppPage.newBooks;
+    } else if (path is AllBooksRoutePath) {
+      _appState.page = AppPage.allBooks;
+    } else if (path is SettingsRoutePath) {
+      _appState.page = AppPage.settings;
+    }
+  }
+
+  @override
+  void dispose() {
+    _appState.dispose();
+    super.dispose();
+  }
+
+  @override
+  AppRoutePath get currentConfiguration {
+    if (_appState.page == AppPage.newBooks) {
+      return NewBooksRoutePath();
+    } else if (_appState.page == AppPage.allBooks) {
+      return AllBooksRoutePath();
+    } else if (_appState.page == AppPage.settings) {
+      return SettingsRoutePath();
+    }
+    return AppRoutePath();
+  }
+
+  void _handleTabSelected(int idx) {
+    if (idx == 0) {
+      _appState.page = AppPage.newBooks;
+    } else {
+      _appState.page = AppPage.allBooks;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    late final Widget innerScreen;
+    if (_appState.page == AppPage.settings) {
+      innerScreen = SettingsScreen();
+    } else {
+      innerScreen = BooksScreen(
+        appPage: _appState.page,
+        onTabSelected: _handleTabSelected,
+      );
+    }
+
+    final bottomBarIndex =
+    _appState.page == AppPage.allBooks || _appState.page == AppPage.newBooks
+        ? 0
+        : 1;
+    return Navigator(
+      key: navigatorKey,
+      pages: [
+        MaterialPage(
```

Comment:
can we separate out the setting and book pages instead of building everything in one page?
for example
```dart
[
  if(_appState.page == AppPage.settings)
    MaterialPage( child: SettingsScreen())
  if (_appState.page == AppPage.newBooks || _appState.page == AppPage.allBooks)
    MaterialPage( child: BooksScreen());
]
```

Reply:
This would also require that the Scaffold is moved above the Navigator I think.